gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageClientGrpcTracingInterceptor.java

```diff
@@ -43,10 +44,14 @@
 public class GoogleCloudStorageClientGrpcTracingInterceptor implements ClientInterceptor {
   private static final GoogleLogger logger = GoogleLogger.forEnclosingClass();
   public static final String IDEMPOTENCY_TOKEN_HEADER = "x-goog-gcs-idempotency-token";
+  public static final String USER_PROJECT_HEADER = "x-goog-user-project";
```

Comment:
Why are we doing this as an interceptor? Why not veneer provide this as a config while creating the client. Even better, they should be able to figure this out from the GCE VM they are running (if it is running from VM), correct?

Reply:
>Why are we doing this as an interceptor?

This is Trace interceptor, here we trace the request for `userProject` filed. If this field is in headers we do log it. It points towards the requesterPays filed being set for gcsbucket in request. 

> Why not veneer provide this as a config while creating the client.

Via this code we are not updating the header to use the provided project but other way round. Java-storage client do provide a way for requester-pays to be set. 

No. It's a userProject header is the one which carries information of `requesterpays` feature of gcs bucket. This has  nothing to with the account/project the VM is in. This feature can even be used from machines which are not part of GCP.